src/topology/semicontinuous.lean

```diff
@@ -0,0 +1,833 @@
+addition, for application to `ereal`. -/
+lemma lower_semicontinuous_within_at.add {f g : α → γ}
+  (hf : lower_semicontinuous_within_at f s x) (hg : lower_semicontinuous_within_at g s x) :
+  lower_semicontinuous_within_at (λ z, f z + g z) s x :=
+hf.add' hg continuous_add.continuous_at
+
+/-- The sum of two lower semicontinuous functions is lower semicontinuous. Formulated with
+`[has_continuous_add]`. The primed version of the lemma uses an explicit continuity assumption on
+addition, for application to `ereal`. -/
+lemma lower_semicontinuous_at.add {f g : α → γ}
+  (hf : lower_semicontinuous_at f x) (hg : lower_semicontinuous_at g x) :
+  lower_semicontinuous_at (λ z, f z + g z) x :=
+hf.add' hg continuous_add.continuous_at
+
+/-- The sum of two lower semicontinuous functions is lower semicontinuous. Formulated with
+`[has_continuous_add]`. The primed version of the lemma uses an explicit continuity assumption on
+addition, for application to `ereal`. -/
+lemma lower_semicontinuous_on.add {f g : α → γ}
+  (hf : lower_semicontinuous_on f s) (hg : lower_semicontinuous_on g s) :
+  lower_semicontinuous_on (λ z, f z + g z) s :=
+hf.add' hg (λ x hx, continuous_add.continuous_at)
+
+/-- The sum of two lower semicontinuous functions is lower semicontinuous. Formulated with
+`[has_continuous_add]`. The primed version of the lemma uses an explicit continuity assumption on
+addition, for application to `ereal`. -/
+lemma lower_semicontinuous.add {f g : α → γ}
+  (hf : lower_semicontinuous f) (hg : lower_semicontinuous g) :
+  lower_semicontinuous (λ z, f z + g z) :=
+hf.add' hg (λ x, continuous_add.continuous_at)
+
+lemma lower_semicontinuous_within_at_sum {f : ι → α → γ} {a : finset ι}
+  (ha : ∀ i ∈ a, lower_semicontinuous_within_at (f i) s x) :
+  lower_semicontinuous_within_at (λ z, (∑ i in a, f i z)) s x :=
+begin
+  classical,
+  induction a using finset.induction_on with i a ia IH generalizing ha,
+  { exact lower_semicontinuous_within_at_const },
+  { simp only [ia, finset.sum_insert, not_false_iff],
+    exact lower_semicontinuous_within_at.add (ha _ (finset.mem_insert_self i a))
+      (IH (λ j ja, ha j (finset.mem_insert_of_mem ja))) }
+end
+
+lemma lower_semicontinuous_at_sum {f : ι → α → γ} {a : finset ι}
+  (ha : ∀ i ∈ a, lower_semicontinuous_at (f i) x) :
+  lower_semicontinuous_at (λ z, (∑ i in a, f i z)) x :=
+begin
+  simp_rw [← lower_semicontinuous_within_at_univ_iff] at *,
+  exact lower_semicontinuous_within_at_sum ha
+end
+
+lemma lower_semicontinuous_on_sum {f : ι → α → γ} {a : finset ι}
+  (ha : ∀ i ∈ a, lower_semicontinuous_on (f i) s) :
+  lower_semicontinuous_on (λ z, (∑ i in a, f i z)) s :=
+λ x hx, lower_semicontinuous_within_at_sum (λ i hi, ha i hi x hx)
+
+lemma lower_semicontinuous_sum {f : ι → α → γ} {a : finset ι}
+  (ha : ∀ i ∈ a, lower_semicontinuous (f i)) :
+  lower_semicontinuous (λ z, (∑ i in a, f i z)) :=
+λ x, lower_semicontinuous_at_sum (λ i hi, ha i hi x)
+
+end
+
+/-! #### Supremum -/
```

Comment:
I'm guessing we'll eventually want the maximum of two lower_semicontinuous functions, as well (perhaps even the `finset.sup` and `finset.sup'`?), but it doesn't need to happen now.

Reply:
Yes, there is a bunch of things that could be added. For now, I've just put the bare minimum I am going to need in the near future (and the PR is already large enough :-)